src/AssetManagement/actions/UniswapV2SwapAction.sol

```diff
@@ -0,0 +1,136 @@
+        actionAssetsData memory _incoming,
+        address[] memory path
+    ) internal {
+        _uniswapV2Swap(_vaultAddress, _outgoing.assetAmounts[0], _incoming.assetAmounts[0], path);
+    }
+
+    function _preCheck(address _vaultAddress, bytes memory _actionSpecificData)
+        internal
+        returns (actionAssetsData memory _outgoing, actionAssetsData memory _incoming, address[] memory path)
+    {
+        /*///////////////////////////////
+                    DECODE
+        ///////////////////////////////*/
+
+        (_outgoing, _incoming, path) = abi.decode(_actionSpecificData, (actionAssetsData, actionAssetsData, address[]));
+
+        require(path.length >= 2, "UV2A_SWAP: _path must be >= 2");
+
+        /*///////////////////////////////
+                    OUTGOING
+        ///////////////////////////////*/
+
+        //TODO Check outgoing assets are not in value more than vault value locked?
+
+        /*///////////////////////////////
+                    INCOMING
+        ///////////////////////////////*/
+
+        //Check if incoming assets are Arcadia whitelisted assets
+        require(
+            IMainRegistry(MAIN_REGISTRY).batchIsWhiteListed(_incoming.assets, _incoming.assetIds),
+            "UV2A_SWAP: Non-allowlisted incoming asset"
+        );
+
+
+        return (_outgoing, _incoming, path);
+    }
+
+    function _postCheck(
+        address _vaultAddress,
+        actionAssetsData memory outgoingAssets_,
+        actionAssetsData memory incomingAssets_
+    ) internal view returns (uint256[] memory incomingAssetAmounts_, uint256[] memory outgoingAssetAmounts_) {
+
+        /*///////////////////////////////
+                    INCOMING
+        ///////////////////////////////*/
+
+        incomingAssetAmounts_ = new uint256[](incomingAssets_.assets.length);
+        for (uint256 i; i < incomingAssets_.assets.length; i++) {
+            incomingAssetAmounts_[i] =
+                IERC20(incomingAssets_.assets[i]).balanceOf(_vaultAddress) - incomingAssets_.preActionBalances[i]; //TODO check overflow?
+
+            // Check incoming assets are as expected
+            require(
+                incomingAssetAmounts_[i] >= incomingAssets_.assetAmounts[i],
+                "UV2A_SWAP: Received incoming asset less than expected"
+            );
+        }
+
+        /*///////////////////////////////
+                    OUTGOING
+        ///////////////////////////////*/
+
+        outgoingAssetAmounts_ = new uint256[](outgoingAssets_.assets.length);
+        for (uint256 i; i < outgoingAssets_.assets.length; i++) {
+            // Calculate the balance change of outgoing assets. Ignore if balance increased.
+            uint256 postActionBalances = IERC20(outgoingAssets_.assets[i]).balanceOf(_vaultAddress);
+            if (postActionBalances < outgoingAssets_.preActionBalances[i]) {
+                outgoingAssetAmounts_[i] = outgoingAssets_.preActionBalances[i] - postActionBalances;
```

Comment:
we know the exact outgoing amount since we use swapExactTokensForTokens

Reply:
Remant from a more generic solution